src/server/pachw/server/pachw.go

```diff
@@ -0,0 +1,154 @@
+package server
+
+import (
+	"context"
+	"path"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/backoff"
+	"github.com/pachyderm/pachyderm/v2/src/internal/dlock"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/middleware/auth"
+	"github.com/pachyderm/pachyderm/v2/src/internal/task"
+	"github.com/pachyderm/pachyderm/v2/src/server/pfs/server"
+)
+
+const (
+	backoffPeriod    = 30 // backoffPeriod is how many seconds the run waits between scale down operations.
+	tickerSeconds    = 5
+	samplesPerMinute = 60 / tickerSeconds
+)
+
+type pachW struct {
+	log *logrus.Logger
+	env Env
+}
+
+func NewController(ctx context.Context, env *Env) {
+	controller := &pachW{
+		env: *env,
+		log: env.Logger,
+	}
+	go controller.run(ctx)
+}
+
+func (p *pachW) run(ctx context.Context) {
+	ctx = auth.AsInternalUser(ctx, "pachw-controller")
+	backoff.RetryUntilCancel(ctx, func() error { //nolint:errcheck
+		eg, ctx := errgroup.WithContext(ctx)
+		eg.Go(func() error {
+			lock := dlock.NewDLock(p.env.EtcdClient, path.Join(p.env.EtcdPrefix, "pachw-controller-lock"))
+			defer func() {
+				if err := lock.Unlock(ctx); err != nil {
+					p.log.Errorf("error unlocking in pachw controller: %v", err)
+				}
+			}()
+			totalInstances := 0
```

Comment:
Could we change this to look more like this:
```
var replicas int
var scaleDownCount int
ticker := time.NewTicker(period)
defer ticker.Stop()
for {
	numTasks := p.countTasks(ctx, []string{server.URLTaskNamespace, server.StorageTaskNamespace})
	scaleDownCount++
	if numTasks > replicas/2 {
		scaleDownCount = 0
	}
	if numTasks > replicas {
		replicas = math.Min(numTasks, p.env.MaxReplicas)
		p.log.Infof("scaling up pachw workers to %d replicas", replicas)
		if err := p.scalePachw(ctx, int32(replicas)); err != nil {
			p.log.Errorf("failed to scale up pachw: %v", err)
		}
	} else if scaleDownCount == scaleDownThreshold {
		replicas = math.Max(replicas / 2, p.env.MinReplicas)
		p.log.Infof("scaling down pachw workers to %d replicas", replicas)
		if err := p.scalePachw(ctx, int32(replicas)); err != nil {
			p.log.Errorf("failed to scale down pachw: %v", err)
		}
	}
	select {
	case <-ticker.C:
	case <-ctx.Done():
		return errors.EnsureStack(ctx.Err())
	}
}
```
I would like this code to be as simple / readable as possible.